### PR TITLE
1450-Add Contrib First documentation.

### PR DESCRIPTION
**Pull request by swirtSJW**
#1450 

----
📚 Documentation preview 📚: https://civicactions-handbook--1451.org.readthedocs.build/en/1451/

### CODE REVIEW

File: common-practices-tools/contribution/contrib-first.md
Comment (dmundra) [inline common-practices-tools/contribution/contrib-first.md]: I would say 'may' receive security coverage since there is a process to do that.

File: common-practices-tools/contribution/contrib-first.md
Comment (dmundra) [inline common-practices-tools/contribution/contrib-first.md]: Can you link to https://drupal.org/civicactions as well to see more of CA's work?

Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: Great idea.

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [inline common-practices-tools/contribution/contrib-first.md]:
```suggestion
Whenever we are building something that could be of use to more than one project or client, we build it as contributed work first.
```

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [inline common-practices-tools/contribution/contrib-first.md]:
```suggestion
-   **Security** - contributing our work to an open source project like Drupal means it may receive security coverage by the Drupal security team and the public.
```

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [inline common-practices-tools/contribution/contrib-first.md:35]:
```suggestion
-   [Vertex AI Search](https://www.drupal.org/project/vertex_ai_search)

See more of [CivicActions Drupal contributions](https://drupal.org/civicactions).
```

**Review by swirtSJW — COMMENTED** [review]
Adjustments made

**Review by dmundra — APPROVED** [review]
Looks great @swirtSJW. Thank you

**Comment by swirtSJW** [comment]
@dmundra Sorry.  Can you take another look?  I added two new sections.

File: common-practices-tools/contribution/contrib-first.md
Comment (dmundra) [inline common-practices-tools/contribution/contrib-first.md]: Maybe link to this page `practice-areas/engineering/drupal/#contribution-to-drupalorg-modules-and-themes` which describes how to do that?

Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: Great suggestion.  This change has been added.

**Review by andyhawks — APPROVED** [review]
👍

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [inline common-practices-tools/contribution/contrib-first.md]:
```suggestion
2. We use the [composer-patches](https://github.com/cweagans/composer-patches) to apply local copies of the contributed patches.
```

File: common-practices-tools/contribution/contrib-first.md
Comment (grugnog) [inline common-practices-tools/contribution/contrib-first.md]: This is pretty Drupal specific language - it would be nice to write this in more general terms (using Drupal as an example, or translating terms etc) as contribution to all the many FOSS projects we use should be our goal.

File: common-practices-tools/contribution/contrib-first.md
Comment (grugnog) [inline common-practices-tools/contribution/contrib-first.md:5]: Not sure quite where this goes, but we should be ensuring we have approval from clients before we contribute code, including talking about security, license and boilerplate the client may need (especially for new projects).
In many cases there are contract terms that cover this and we have or can get blanket approvals for various types of contribution.

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: Good point.  It does make a fare amount of assumptions related to Drupal.  I wanted this to be the reference specifically for Drupal that I could point to, mainly because the whole issue queue and contrib credit does not really exist elsewhere.  What if doubled down on making this one section Drupal specific and then added another section that was more agnostic for other FOSS contributions?

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md:5]: Hmm maybe a sentence or two in the top section.  I'll take a whack at it.

File: common-practices-tools/contribution/contrib-first.md
Comment (grugnog) [reply common-practices-tools/contribution/contrib-first.md]: I think that would work!

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md:5]: I added this concept to the top.

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (dmundra) [inline practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: I think you don't need to duplicate the header on this page. I believe the solution to any linting errors is to add the frontmatter title. Example:

```
---
title: Drupal Contrib First module development
---
```

I also noticed you want to do that same in the 'contrib-first' page as well. Also small typo in the title.

Comment (dmundra) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Or the solution is to use single '#' for page title h1 and not two '#'

File: practice-areas/engineering/drupal/drupal-developer-tips-for-getting-the-most-out-of-open-source.md
Comment (dmundra) [inline practice-areas/engineering/drupal/drupal-developer-tips-for-getting-the-most-out-of-open-source.md]: Add a space between '->Community Participation'

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (dmundra) [inline practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Might need to ignore the indentation problem. I am going to look for an example of indented lists.

Comment (dmundra) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Here is one https://github.com/CivicActions/guidebook/blob/master/practice-areas/project-management/project-management-checklists.md?plain=1#L41

File: common-practices-tools/contribution/contrib-first.md
Comment (dmundra) [inline common-practices-tools/contribution/contrib-first.md:5]: @swirtSJW I forgot that titles should have frontmatter as well, e.g. https://github.com/CivicActions/guidebook/blob/master/practice-areas/project-management/project-management-checklists.md?plain=1#L1-L3

File: common-practices-tools/contribution/contrib-first.md
Comment (dmundra) [inline common-practices-tools/contribution/contrib-first.md]: Can we add some non Drupal ones also change to 'Example of open source repos ...) Examples:

* https://github.com/GetDKAN/cmsds-open-data-components (maybe link DKAN in general)
* https://github.com/drydockcloud
* This guidebook?
* https://github.com/CivicActions/accessibility
* https://github.com/GSA/openacr

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (dmundra) [inline practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Add link to https://www.drupal.org/drupal-security-team/security-advisory-process-and-permissions-policy

**Review by dmundra — CHANGES_REQUESTED** [review]
@swirtSJW sorry I noticed some things that need updating related to linting and markdown (frontmatter comments). Also remembered some useful additions like non-Drupal examples and links.

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (dmundra) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Double title issue seen here https://civicactions-handbook--1451.org.readthedocs.build/en/1451/practice-areas/engineering/drupal/drupal-contrib-first-module-development/

File: common-practices-tools/contribution/contrib-first.md
Comment (grugnog) [inline common-practices-tools/contribution/contrib-first.md]: Typo - should be CivicActions

Comment (grugnog) [inline common-practices-tools/contribution/contrib-first.md]: What about broadening this to non-federal, non-CivicActions users - e.g. "CivicActions other clients and the public at large can benefit"

File: common-practices-tools/contribution/contrib-first.md
Comment (grugnog) [inline common-practices-tools/contribution/contrib-first.md]: Security coverage (at least in terms of Drupal) doesn't actually lead to detection of any new security issues - it's focus is really only on coordinating fixes (completed by project maintainers) for issues that are reported by external users and then notifying other users.
This is a benefit of course, but I am not sure it necessarily adds security on it's own. I would consider starting out with the "[many eyes](https://en.wikipedia.org/wiki/Linus%27s_law)" argument. Essentially, by having many users it's more likely that security vulnerabilities will be detected over time (if it remains private, this is much less likely). Combining this with vulnerability reporting/notification infrastructure (like Drupal, but NVD is also a great example) you get the benefit of a more secure site with minimal additional effort.

**Review by grugnog — DISMISSED** [review]

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: CivicActions Accessibility and our guidebook does not seem contrib first but maybe my interpretation is flawed.  

1. It seems quite specific to CivicActions rather than something anyone else can use unaltered.
2. It seems like it was not built for a client out of a client need
3. It seems more like it was past work, amalgamated into new public repos.

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: I adjusted this to include that.

File: common-practices-tools/contribution/contrib-first.md
Comment (dmundra) [reply common-practices-tools/contribution/contrib-first.md]: Ah I think both of these warrant #ca-stories stories ;)

Guidebook used to be a full fork of https://github.com/18F/handbook and on the readme we have this line:

> We'd like to thank the folks at 18F for [their handbook](https://github.com/18F/handbook), from which this documentation was forked. We welcome you to fork our documentation, or use the 18F handbook, which is licensed as CC0.

I also believe folks outside of CA have forked our guidebook.

For accessibility, that was also a full fork of https://github.com/18F/uswds-jekyll where we want to create not just the accessibility site but show how folks can quickly spin up sites on GitHub. Maybe this is a better example: https://github.com/CivicActions/ca-jekyll-vanilla

So yes it is a stretch but they were build from opensource projects to show how someone else can. Folks are welcome then to use our examples to build their own. Meets 1., not 2., and yes it is 3.

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: Nice.   Done.

Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: Fixed

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (swirtSJW) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Added.

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md]: I added 3 out of the 5.

File: common-practices-tools/contribution/contrib-first.md
Comment (swirtSJW) [reply common-practices-tools/contribution/contrib-first.md:5]: Thank you.  Fixed.

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (swirtSJW) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Fixed.

File: practice-areas/engineering/drupal/drupal-developer-tips-for-getting-the-most-out-of-open-source.md
Comment (swirtSJW) [reply practice-areas/engineering/drupal/drupal-developer-tips-for-getting-the-most-out-of-open-source.md]: Fixed

File: practice-areas/engineering/drupal/drupal-contrib-first-module-development.md
Comment (swirtSJW) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Fixed

Comment (dmundra) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: Cool, still seeing an extra header so I think that extra one can be dropped.

Comment (swirtSJW) [reply practice-areas/engineering/drupal/drupal-contrib-first-module-development.md]: late push fixed it :)

**Review by dmundra — APPROVED** [review]
Thank you @swirtSJW

**Pull request merged**
